### ISSUE TITLE
"Add Album to Queue" and "Play Album Next" don't work in Chromium build

### ISSUE THREAD

**Issue by markstos** [tiliado/nuvola-app-google-play-music#24]
Hello, I'm testing the Chromium backend. Version 6, on Nuvola Apps 4.9.96.

I'm looking at my Recently Played page.  I click on an album and select from the "dot dot dot" menu the option "Play Album Next" or "Add Album to Queue". Then I look in the queue in the lower right and it's empty. 

To confirm this is a bug, I repeated the same steps in Firefox and confirmed albums were added to the queue. 

I tried again with the DevTools console open. I saw no errors in the "Console" tab, and no failed requests in the "Network" tab.

These features also works in the WebKit version.

**Comment by fenryxo** [comment]
Hello @markstos. I cannot reproduce your issue under Xorg. Are your running under Wayland? There has been something similar reported at https://github.com/tiliado/nuvolaruntime/issues/372#issuecomment-355214224

**Comment by markstos** [comment]
Yes, ` echo $XDG_SESSION_TYPE` reports "wayland", so it appears be an exact dupe of that issue. (Although I don't know which bug queue is most appropriate for the issue).

**Comment by fenryxo** [comment]
Hello @bartlibert. The fact that the issue still occurs with Chromium 65 is enough for now. I recommend using Xorg until it is resolved.

**Comment by fenryxo** [comment]
It seems to work, but very erratically, i.e., I have to try multiple times until the menu entry activation succeeds. Can anybody verify my observation with the lastest Nuvola as well as with `cefsimple` from [the latest Linux CEF Sample Application tarball](http://opensource.spotify.com/cefbuilds/index.html)?

* When I click a three-dot button, it always opens menu successfully.
* When I click a menu item, the menu is always closed regardless whether the corresponding action was performed or not.
* When I click a menu item, it usually does not work, but sometimes it does, i.e., I must keep trying more times to succeed eventually.

**Comment by fenryxo** [comment]
Note to self: Check whether mouse events are emitted in the correct order.

**Comment by bartlibert** [comment]
I can confirm following things with the latest nuvola build:

- When I click a three-dot button, it always opens menu sccessfully.
- When I click a menu item, the menu is always closed regardless whether the corresponding action was performed or not.

However, no matter how much I try, the action is never executed on my system.
The cefsimple test application always crashes with " Unable to open X display", so I cannot verify that.

**Comment by fenryxo** [comment]
That cefsimple crash is weird. Do other X apps work? E.g. xclock, xcalc?

**Comment by fenryxo** [comment]
Another funny thing. When I click the three-dot button and then on the first menu item without moving a cursor, the menu is closed an then reopened again. It looks like the click event was ignored by menu item, but captured by the three-dot button.

**Comment by bartlibert** [comment]
I tried again with cefsimple (downloaded again etc..) and now it works. The results are a bit different than in nuvola:
- Clicking the menu sometimes fails (a second click is needed to show it) -> in nuvola it always works
- Clicking an action sometimes (rarely) works, but most of the time it does not

**Comment by fenryxo** [comment]
I was about to investigate this issue with Chromium 66 the last week but got SSD disk failure during the build process :-/ I need to make space on another disk.

**Comment by fenryxo** [comment]
[There is ongoing work on native Wayland support in Chromium](https://bugs.chromium.org/p/chromium/issues/detail?id=578890)

**Comment by markstos** [comment]
@fenryxo Thanks for the link. I thought ChromeOS already used Wayland, so Chromium was already running on Wayland, but I must have been mistaken. I'm glad to see the progress, though!

**Comment by fenryxo** [comment]
Hello @markstos & @bartlibert. Could you test with the latest Nuvola/master with  `VALACEF_DEFAULT_RENDERING_MODE=offscreen`? E.g. `VALACEF_DEFAULT_RENDERING_MODE=offscreen flatpak run eu.tiliado.NuvolaAppGooglePlayMusic//master -D`? You may need to run `flatpak install nuvola eu.tiliado.NuvolaAppGooglePlayMusic//master` first.

**Comment by markstos** [comment]
@fenryxo I tested using  633946737f38.

Using `VALACEF_DEFAULT_RENDERING_MODE=offscreen`, I confirmed that "Add Album to Queue now works".

If I quit and run it again without this environment variable set, the feature is broken.

**Comment by bartlibert** [comment]
I can confirm it is working with and working without the environment variable.

**Comment by fenryxo** [comment]
Great, so the offscreen rendering seems to be a promising way to go as it solves three bugs. Unfortunately, the memory & CPU usage suffers as Nuvola needs to keep the pixel buffer around (~40 MB for Full HD) and does the final drawing with CPU instead of GPU ([Poster circle](https://webkit.org/blog-files/3d-transforms/poster-circle.html) gives 23 fps instead of 60 fps). That needs to be optimized. More at tiliado/valacef#11
